### ISSUE TITLE
Variable length data in Fuel

### ISSUE THREAD

**Issue by rizar** [mila-udem/fuel#136]
Guys, I would like to know how you envisage handling datasets with variable-length examples in Fuel, e.g. speech, handwriting, music data (I do not mention text, because we sort of support text files). Do you think they should be also converted to HDF5? There is some sort of support of variable length data in HDF5, but I haven't tried it so far. In both machine translation and speech projects I participated we used tables with concatenated sequences and offset-length tables. This works, but it is quite different from what `HDF5Dataset` currently expects.

I am asking without doing too much of prior research, because I would like to know if you have already thought it through.

**Comment by bartvm** [comment]
https://github.com/mila-udem/fuel/pull/133#issuecomment-104787991

**Comment by rizar** [comment]
Cool, thanks! I guess a separate issue might be a better place for such a principal discussion.

I think we should be careful not to grow `HDF5Dataset` too large, but I also agree that reshapes should be done before transformers. A way to go could be a `VariableLengthHDF5Dataset` that inherits from `HDF5Dataset` and does reshapes and other stuff.

**Comment by vdumoulin** [comment]
The reshaping could be done automatically, I think: `h5py` has a 'dimension scale' feature which could allow us to tag a dimension with the information required for reshaping. If `H5PYDataset` sees such dimension scales attached to a dataset, it would perform the required reshaping in `get_data`.

**Comment by vdumoulin** [comment]
I've started a pull request for this new feature: #141.
